File: firefly-example/src/main/java/com/firefly/example/reactive/coffee/store/router/impl/sys/GlobalHandler.java
package com.firefly.example.reactive.coffee.store.router.impl.sys;

import com.firefly.$;
import com.firefly.annotation.Component;
import com.firefly.codec.http2.model.HttpStatus;
import com.firefly.example.reactive.coffee.store.vo.Response;
import com.firefly.example.reactive.coffee.store.vo.ResponseStatus;
import com.firefly.server.http2.router.Handler;
import com.firefly.server.http2.router.RoutingContext;
import com.firefly.utils.log.slf4j.ext.LazyLogger;
import reactor.core.publisher.Mono;

import java.time.Duration;

/**
 * @author Pengtao Qiu
 */
@Component("globalHandler")
public class GlobalHandler implements Handler {

    private static final LazyLogger logger = LazyLogger.create();

    @Override
    public void handle(RoutingContext ctx) {
        ctx.getResponse().setAsynchronous(true);
        Mono.fromFuture(ctx.nextFuture())
            .timeout(Duration.ofSeconds(1))
            .subscribe(ret -> ctx.end(), ex -> {
                if (ex instanceof IllegalArgumentException) {
                    renderIllegalArgumentException(ctx, (IllegalArgumentException) ex);
                } else if (ex.getCause() instanceof IllegalArgumentException) {
                    renderIllegalArgumentException(ctx, (IllegalArgumentException) ex.getCause());
                } else if (ex instanceof IllegalStateException) {
                    renderIllegalStateException(ctx, (IllegalStateException) ex);
                } else if (ex.getCause() instanceof IllegalStateException) {
                    renderIllegalStateException(ctx, (IllegalStateException) ex.getCause());
                } else {
                    ctx.setStatus(HttpStatus.INTERNAL_SERVER_ERROR_500);
                    logger.error(() -> "Server exception. " + ctx.getURI(), ex);
                    Response<Boolean> response = new Response<>();
                    response.setStatus(ResponseStatus.SERVER_ERROR.getValue());
                    response.setMessage("server exception, " + ex.getMessage());
                    ctx.writeJson(response).end();
                }
            });
    }

    private void renderIllegalArgumentException(RoutingContext ctx, IllegalArgumentException ex) {
        ctx.setStatus(HttpStatus.OK_200);
        logger.info(() -> $.string.replace("Request error. [{}], [{}]", ctx.getURI(), ex.getMessage()));
        Response<Boolean> response = new Response<>();
        response.setStatus(ResponseStatus.ARGUMENT_ERROR.getValue());
        response.setMessage(ex.getMessage());
        ctx.writeJson(response).end();
    }

    private void renderIllegalStateException(RoutingContext ctx, IllegalStateException ex) {
        ctx.setStatus(HttpStatus.OK_200);
        logger.info(() -> $.string.replace("Request error. [{}], [{}]", ctx.getURI(), ex.getMessage()));
        Response<Boolean> response = new Response<>();
        response.setStatus(ResponseStatus.SERVER_ERROR.getValue());
        response.setMessage(ex.getMessage());
        ctx.writeJson(response).end();
    }
}

File: firefly-example/src/main/kotlin/com/firefly/example/kotlin/coffee/store/router/impl/sys/GlobalHandler.kt
package com.firefly.example.kotlin.coffee.store.router.impl.sys

import com.firefly.annotation.Component
import com.firefly.codec.http2.model.HttpStatus
import com.firefly.example.kotlin.coffee.store.vo.Response
import com.firefly.example.kotlin.coffee.store.vo.ResponseStatus
import com.firefly.kotlin.ext.http.AsyncHandler
import com.firefly.kotlin.ext.http.asyncNext
import com.firefly.kotlin.ext.log.KtLogger
import com.firefly.server.http2.router.RoutingContext
import java.lang.IllegalArgumentException

/**
 * @author Pengtao Qiu
 */
@Component("globalHandler")
class GlobalHandler : AsyncHandler {

    private val log = KtLogger.getLogger { }

    override suspend fun handle(ctx: RoutingContext) {
        log.info("request start -> ${ctx.uri}")
        ctx.asyncNext<Unit>(succeeded = {
            log.info("request end -> ${ctx.uri}")
            ctx.end()
        }, failed = {
            log.error("request error -> ${ctx.uri}", it)
            when (it) {
                is IllegalArgumentException -> {
                    ctx.setStatus(HttpStatus.OK_200)
                    ctx.writeJson(
                        Response(
                            ResponseStatus.ARGUMENT_ERROR.value,
                            it.message ?: ResponseStatus.ARGUMENT_ERROR.description,
                            false
                                )
                                 ).end()
                }
                is IllegalStateException -> {
                    ctx.setStatus(HttpStatus.OK_200)
                    ctx.writeJson(
                        Response(
                            ResponseStatus.SERVER_ERROR.value,
                            it.message ?: ResponseStatus.SERVER_ERROR.description,
                            false
                                )
                                 ).end()
                }
                else -> {
                    log.error("server exception", it)
                    ctx.setStatus(HttpStatus.INTERNAL_SERVER_ERROR_500)
                    ctx.writeJson(
                        Response(
                            ResponseStatus.SERVER_ERROR.value,
                            it?.message ?: ResponseStatus.SERVER_ERROR.description,
                            false
                                )
                                 ).end()
                }
            }
        })
    }

}